### PR TITLE
2199 fiddle

### LINKED ISSUE #2199: Support the use of ”Fiddle” files
In this episode of the Clojure Design Podcast https://clojuredesign.club/episode/014-fiddle-with-the-repl/ they talk about using separate files to put code experiments and such in. It's a bit like Rich Comment Forms in a separate file. They call them fiddle files. It's not too different from how the REPL Window works, except that there is no need for a prompt, because a fiddle file typically only serves one namespace. (At least that's one way to do it.)

The fiddle files are supposed to go in some `dev` source path, so that they do not get part of the packaged and shipped application. Calva could make it easy to switch back and forth between the fiddle file for a namespace and the implementation of that namespace. A bit like we do with test files. (Or maybe it is clojure-lsp doing it these days.) This would take a bit of configuration from the user, I think. So that we know which folder to treat as the ”mirror”. I think we might need to know things about the classpath as well. Maybe clojure-lsp can give us that.

Fiddle files could also be placed side-by-side with the implementation namespace they server. I sometimes do this, and name the files with `.calva-repl` as their extension, which gets treated a bit like `.cljc` files, because that's how the REPL Window is treated. We could support some more file extension to be treated like this. Like `.fiddle`, and `.rich`. Switching to and from these files is easier to implement.

### PR DESCRIPTION
**Pull request by PEZ**
## What has changed?

For some context see: [Episode 014: Fiddle with the REPL](https://clojuredesign.club/episode/014-fiddle-with-the-repl/)

* Adding commands for switching between “fiddle” files and their corresponding source files, and for evaluating fiddle files while the corresponding source remains active.
* Making `.fiddle` files handled as Clojure files
* Adding settings for specifying type of fiddle file
  * default to sibling `.fiddle` files
  * can be set to create a sibling source tree somewhere in the project. E.g. `./env/dev/fiddles` and then fiddles will be create there with the same file extension as their source files.

WIP: Further description TBD, as well as docs. Also, some testing on Windows is probably good, because paths...

* Fixes #2199 

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Made sure there is an issue registered with a clear problem statement that this PR addresses, (created the issue if it was not present).
    - [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
    - [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
        - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -162,6 +163,34 @@
         "type": "object",
         "title": "Calva",
         "properties": {
+          "calva.fiddleFilePaths": {
+            "type": "array",
+            "markdownDescription": "An array of `source` and `fiddle` files files root paths, relative to the project root. This is used by the command **Calva: Open Fiddle File**. E.g.:\n\n```json\n\"calva.fiddleFilePaths\": [\n  {\n    \"source\": [\"src\"],\n    \"fiddle\": [\"dev\", \"fiddles\"]\n  }\n]\n```\n\nThe default is `null`, which will make Calva associate files with a `.fiddle` file extension as the fiddle file for a Clojure file. E.g. a file `src/foo/bar/baz_main.cljc`, will be associated with a file named `src/foo/bar/baz_main.fiddle` as its fiddle file. See [calva.io/fiddle](https://calva.io/fiddle/) for more info.",
```
Comment (bpringe) [inline package.json]:
```suggestion
            "markdownDescription": "An array of `source` and `fiddle` files root paths, relative to the project root. This is used by the command **Calva: Open Fiddle File**. E.g.:\n\n```json\n\"calva.fiddleFilePaths\": [\n  {\n    \"source\": [\"src\"],\n    \"fiddle\": [\"dev\", \"fiddles\"]\n  }\n]\n```\n\nThe default is `null`, which will make Calva associate files with a `.fiddle` file extension as the fiddle file for a Clojure file. E.g. a file `src/foo/bar/baz_main.cljc`, will have an associated fiddle file named `src/foo/bar/baz_main.fiddle`. See [calva.io/fiddle](https://calva.io/fiddle/) for more info.",
```

**Comment by bpringe** [comment]
I don't see myself using this feature, but I may just not be the intended audience. I do use something like a fiddle file in a work project that's specific to me, but it's not ns-specific, and I like it that way.

I have not listened to that linked podcast episode, though.

**Comment by PEZ** [comment]
It'll be quite interesting to learn if someone else but me will be using this.

So, your fiddle file is more like a scrap area, I take it? With some little extra work this feature could be configured to support that too. But it depends on how you use it, if it would be of any value, of course. Do you use rich comments in that file?

**Comment by bpringe** [comment]
> So, your fiddle file is more like a scrap area, I take it?

Yes. I use a personal dev namespace as mentioned in the Polylith tool documentation: https://polylith.gitbook.io/poly/architecture/development

> One way of structuring the code is to give all developers their own namespace under the dev top namespace. Let's follow that pattern and create the namespace dev.lisa.

This personal dev file is committed so I have the history of it and have easy access to it across different machines.

> With some little extra work this feature could be configured to support that too. But it depends on how you use it, if it would be of any value, of course.

I'm not sure what type of support this feature could add to that.

> Do you use rich comments in that file?

Yes, most of the code in that file is in a rich comment. And there may be a few forms at the top level.

**Comment by PEZ** [comment]
Cool. In a hobby Polylith project of mine I am using a personal dev namespace root, also inspired from that same documentation page. So the project namespaces are mirrored into `dev.pez/components/...` and such.

> > With some little extra work this feature could be configured to support that too. But it depends on how you use it, if it would be of any value, of course.
> 
> I'm not sure what type of support this feature could add to that.

Since most of your code in that file is in rich comments. You could have use for the command that evaluates the fiddle file without leaving the implementation file.

I'm placing test code for what I am currently working with, last in the fiddle file, _outside_ rich comments. When I execute the **Evaluate Fiddle File** command, my test code is evaluated. That means I don't need to move the cursor back and forth to and from a rich comment to test things. When it works as I want, I either throw away the test code, or tuck it in a rich comment, leaving it in the fiddle file, or moving it to the implementation file, if I think it should be there.

**Comment by bpringe** [comment]
> Since most of your code in that file is in rich comments. You could have use for the command that evaluates the fiddle file without leaving the implementation file.

I think I am not following here why I would want to evaluate the fiddle file without leaving the implementation file.

> I'm placing test code for what I am currently working with, last in the fiddle file, outside rich comments. When I execute the Evaluate Fiddle File command, my test code is evaluated. That means I don't need to move the cursor back and forth to and from a rich comment to test things. When it works as I want, I either throw away the test code, or tuck it in a rich comment, leaving it in the fiddle file, or moving it to the implementation file, if I think it should be there.

I don't really follow how that workflow is any different than what I currently do (or could do) in any file, but maybe I'm not seeing the whole picture. Maybe a video of the workflow at some point would help.

**Comment by PEZ** [comment]
> Maybe a video of the workflow at some point would help.

I will definitely create a video with this. But in waiting for this, please have a look at the docs I just added https://2199-fiddle--calva-docs.netlify.app/fiddle-files/ and see if that clarifies anything. Or even if it doesn't clarify much more than you already know, please check for all spelling mistakes and my broken English and such. 🙏 

(In the process I also setup so that we get branch docs built and published 🚀):

**Pull request merged**